[VPlan] Handle early exit loops with non-dereferenceable loads in latch

lukel97 opened the pull request:
Stacked on #176143 

If we have an early exit loop with non-dereferenceable loads after the exit, we currently bail:

    int z;
    for (int i = 0; i < N; i++) {
      if (x[i])
        break;
      z = y[i];
    }

If the early exit block dominates the block containing these loads, we can predicate them with a mask like

    for (int i = 0; i < N/VF; i++) {
      c[0..VF] = x[i..i+VF]
      z[0..VF] = y[i..i+VF], mask=(icmp ult step-vector, (first-active-lane c))
      if (anyof(c))
        break;
    }

This patch handles by creating a mask of lanes past the exiting lane, and branching to the latch on it. It reuses the same infrastructure as tail folding in #176143 to do so.

This makes way for supporting more types of loops, as we could also support stores/divs etc. as long as all exiting blocks dominate them. See the note in canUncountableExitConditionLoadBeMoved as to why we can't predicate stores when they're before the exiting block.

In order to actually test the changes from this, this PR allows non-dereferenceable loads that are properly dominated by the exiting block in LoopVectorizationLegality. But in practice, something else usually transforms these loops to be multiple-entry which prevents them from being vectorized.

lukel97 commented:
> I wonder if it’d be possible to make this a post-VPlanPredicator transform and leave VPlanPredicator unchanged, which is similar to what we do for EVL recipes.

I think that would leave us with an incorrect VPlan temporarily which would be nice to avoid. But we might be able to do it as a pre-VPlanPredicator transform, I'll try exploring this. 

> Also as noted in https://github.com/annamthomas/work-notes/blob/main/MultiDataDependentExitVectorization.rst, have you also considered to predicate the loop body in addition to successors? We may need to evaluate the hoistability and hoist the early‑exit condition to the header to ensure it dominates all operations.

I presume we can hoist early exit conditions in an earlier loop canonicalization pass? Do we do this already somewhere?

arcbbb reviewed (commented):
LGTM! Pending approval from others.

arcbbb commented:
> This makes way for supporting more types of loops, as we could also support stores/divs etc. as long as the exiting block dominates them. See the note in canUncountableExitConditionLoadBeMoved as to why we can't predicate stores when they're before the exiting block.
> 

The description mentions `canUncountableExitConditionLoadBeMoved`, but I can't find it in the code.

lukel97 committed:
[VPlan] Move tail folding out of VPlanPredicator. NFC

Currently the logic for introducing a header mask and predicating the vector loop region is done inside introduceMasksAndLinearize.

This splits the tail folding part out into an individual VPlan transform so that VPlanPredicator.cpp doesn't need to worry about tail folding, which seemed to be a temporary measure according to a comment in VPlanTransforms.h.

To perform tail folding independently, this splits the "body" of the vector loop region between the phis in the header and the branch + iv increment in the latch:

Before:

    +-------------------------------------------+
    |%iv = ...                                  |
    |...                                        |
    |%iv.next = add %iv, vfxuf                  |
    |branch-on-count %iv.next, vector-trip-count|
    +-------------------------------------------+

After:

    +-------------------------------------------+
    |%iv = ...                                  |
    |%wide.iv = widen-canonical-iv ...          |
    |%header-mask = icmp ult %wide.iv, BTC      |---+
    |branch-on-cond %header-mask                |   |
    +-------------------------------------------+   |
                         |                          |
                         v                          |
    +-------------------------------------------+   |
    |...                                        |   |
    +-------------------------------------------+   |
                         |                          |
                         v                          |
    +-------------------------------------------+   |
    |%iv.next = add %iv, vfxuf                  |<--+
    |branch-on-count %iv.next, vector-trip-count|
    +-------------------------------------------+

The motivation for this is to align tail folding predication with early-exit predication in #172454. This will also allow us to directly emit an EVL based header mask, instead of having to match + transform the existing header mask in addExplicitVectorLength.

lukel97 committed:
Don't use branch-on-cond, explicitly pass mask in densemap to VPPredicator

This preserves SSA

lukel97 committed:
Don't pass state, use VPInstruction::PredicateSuccessors

lukel97 committed:
Revert "Don't pass state, use VPInstruction::PredicateSuccessors"

This reverts commit 81332b3fa28e95819fe9d67310a45409b15a833e.

lukel97 committed:
Revert "Don't use branch-on-cond, explicitly pass mask in densemap to VPPredicator"

This reverts commit ed18c28e87494515bf398919b68fbb0a0527123d.

lukel97 committed:
Maintain SSA

lukel97 committed:
Remove unused VPTypeAnalysis

lukel97 committed:
Assert no other recipes are in new latch

lukel97 committed:
Iterate through defs instead of uses

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/move-foldTailByMasking

lukel97 committed:
Merge branch 'main' into loop-vectorize/move-foldTailByMasking

lukel97 committed:
Add asserts

lukel97 committed:
Undo convertPhisToBlends change

lukel97 committed:
Update diagram with phis

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/move-foldTailByMasking

lukel97 committed:
Check Header instead of VPBB, move assert closer

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/move-foldTailByMasking

lukel97 committed:
Update TODO

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/move-foldTailByMasking

lukel97 committed:
Add VPlan tests

lukel97 committed:
Move to just after createLoopRegions, precommit for vplan test diff

lukel97 committed:
VPlan test diff

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/move-foldTailByMasking

lukel97 committed:
Use VPTypeAnalysis, rework to collect users separately and then add phis

lukel97 committed:
Add test cases for header with only phis, and live-out defined by phi

lukel97 committed:
Explicitly check for VPWidenInductionRecipe

lukel97 committed:
Fix replacing users inside the loop

This silently just happened to work because these phis are just replaced with the same original value.

lukel97 committed:
Use MapVector and cast VPHeaderPHIRecipe

lukel97 committed:
Merge branch 'main' into loop-vectorize/move-foldTailByMasking

lukel97 committed:
[VPlan] Split the latch in a separate transform. NFC

lukel97 committed:
[VPlan] Handle early exit loops with non-dereferenceable loads in latch
